File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        n_str = str(n)
        hash_set = set()
        while True:
            intermediate = 0
            for i in n_str:
                intermediate += int(i) * int(i)
            if intermediate == 1:
                return True
            if intermediate not in hash_set:
                hash_set.add(intermediate)
            else:
                return False
            n_str = str(intermediate)
            

File: 0202-happy-number/0202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> seen;
        while (n != 1 && seen.find(n) == seen.end()) {
            seen.insert(n);
            n = getNext(n);
        } 
        return n == 1;
    }
private:
    int getNext(int x) {
        int res = 0;
        while (x > 0) {
            int digit = x % 10;
            x /= 10;
            res += digit * digit;
        }
        return res;
    }
};